### PR TITLE
[CELEBORN-234] Implement buffer stream.

### PR DESCRIPTION
**Pull request by FMX**
### What changes were proposed in this pull request?
This pr enable Celeborn to support buffer stream and credit-based data flow.

### Why are the changes needed?
To support buffer stream.

### Does this PR introduce _any_ user-facing change?
NO.

### How was this patch tested?
Flink Integration test.

### CODE REVIEW

File: common/src/main/java/org/apache/celeborn/common/network/protocol/ReadData.java [hunk 1]
```diff
@@ -14,48 +14,39 @@
...
   @Override
   public int encodedLength() {
-    return 8 + 4 + 4 + 8 + 4 + buf.readableBytes();
+    return 8 + 4 + 8;
   }
 
   @Override
-  public void encode(ByteBuf buf) {
+  public void encode(io.netty.buffer.ByteBuf buf) {
     buf.writeLong(streamId);
```
Comment (zhongqiangczq) [inline common/src/main/java/org/apache/celeborn/common/network/protocol/ReadData.java:47]: io.netty.buffer.ByteBuf prefix is unnecessary

File: common/src/main/java/org/apache/celeborn/common/network/protocol/ReadData.java [hunk 2]
```diff
@@ -14,48 +14,39 @@
...
 
   @Override
   public int encodedLength() {
-    return 8 + 4 + 4 + 8 + 4 + buf.readableBytes();
+    return 8 + 4 + 8;
   }
 
   @Override
-  public void encode(ByteBuf buf) {
+  public void encode(io.netty.buffer.ByteBuf buf) {
```
Comment (zhongqiangczq) [inline common/src/main/java/org/apache/celeborn/common/network/protocol/ReadData.java:46]: "io.netty.buffer.ByteBuf" prefix  is unnecessary

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 3]
```diff
@@ -27,6 +27,7 @@ import scala.util.Try
 
 import org.apache.hadoop.security.UserGroupInformation
 
+import org.apache.celeborn.common.CelebornConf.{PARTITION_READ_BUFFERS_MAX, PARTITION_READ_BUFFERS_MIN}
```
Comment (zhongqiangczq) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:30]: the import is unnecessary

File: worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/MapPartitionFileWriter.java [hunk 4]
```diff
@@ -199,6 +199,7 @@ public synchronized void destroy(IOException ioException) {
   public void pushDataHandShake(int numReducePartitions, int bufferSize) {
     this.numReducePartitions = numReducePartitions;
     fileInfo.setBufferSize(bufferSize);
+    fileInfo.setNumReducerPartitions(numReducePartitions);
```
Comment (zhongqiangczq) [inline worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/MapPartitionFileWriter.java:202]: "fileInfo.setNumReducerPartitions(numReducePartitions);" will be added by another pr

File: common/src/main/scala/org/apache/celeborn/common/util/Utils.scala [hunk 5]
```diff
@@ -994,4 +996,64 @@ object Utils extends Logging {
...
+      logError(s"Incorrect buffer header, buffer length: ${bufferLength}.")
+      throw new RuntimeException(s"File ${fileChannel} is corrupted")
+    }
+    // attach header buffer to data buffer.
+    buffer.writeBytes(header)
+    readBuffer(fileChannel, buffer, bufferLength)
+    bufferLength + headerSize
+  }
+
+  def getShortFormattedFileName(fileInfo: FileInfo): String = {
```
Comment (zhongqiangczq) [inline common/src/main/scala/org/apache/celeborn/common/util/Utils.scala:1052]: getShortFormattedFileName can be removed. another pr will removed this,

File: common/src/main/scala/org/apache/celeborn/common/util/Utils.scala [hunk 6]
```diff
@@ -994,4 +996,64 @@ object Utils extends Logging {
...
+  def readBuffer(
+      fileChannel: FileChannel,
+      header: ByteBuffer,
+      buffer: ByteBuf,
+      headerSize: Int): Int = {
+    readBuffer(fileChannel, header, headerSize)
+    val bufferLength = header.getInt(12)
+    if (bufferLength <= 0 || bufferLength > buffer.capacity) {
+      logError(s"Incorrect buffer header, buffer length: ${bufferLength}.")
+      throw new RuntimeException(s"File ${fileChannel} is corrupted")
```
Comment (zhongqiangczq) [inline common/src/main/scala/org/apache/celeborn/common/util/Utils.scala:1044]: by fileChannel, we can't directly get the filename if log errors. can readbuffer add a parameter filename ?

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 7]
```diff
@@ -2948,4 +2954,28 @@ object CelebornConf extends Logging {
       .doc("The size of network buffers required per result partition. The minimum valid value is 8M. Usually, several hundreds of megabytes memory is enough for large scale batch jobs.")
       .stringConf
       .createWithDefault("64m")
+
+  val PARTITION_READ_BUFFERS_MIN: ConfigEntry[Int] =
+    buildConf("celeborn.partition.initial.readBufferMin")
```
Comment (zhongqiangczq) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:2959]: can readBufferMin be replaced readBuffersMin?

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 8]
```diff
@@ -757,6 +758,11 @@ class CelebornConf(loadDefaults: Boolean) extends Cloneable with Logging with Se
     get(WORKER_DIRECT_MEMORY_RATIO_FOR_SHUFFLE_STORAGE)
   def memoryPerResultPartition: String = get(MEMORY_PER_RESULT_PARTITION)
 
+  def partitionReadBuffersMin: Int = get(PARTITION_READ_BUFFERS_MIN)
+
+  def partitionReadBuffersMax: Int = get(PARTITION_READ_BUFFERS_MAX)
+  def bufferStreamThreadsPerMountpoint = get(BUFFERSTREAM_THREADS_PER_MOUNTPOINT)
```
Comment (zhongqiangczq) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:764]: def bufferStreamThreadsPerMountpoint => def bufferStreamThreadsPerMountpoint: Int

**Review by zhongqiangczq — APPROVED** [review]

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/FetchHandler.scala [hunk 9]
```diff
@@ -86,6 +89,11 @@ class FetchHandler(val conf: TransportConf) extends BaseMessageHandler with Logg
 
   def handleOpenStream(client: TransportClient, request: RpcRequest): Unit = {
     val msg = Message.decode(request.body().nioByteBuffer())
+    if (msg.`type`() == Type.BACKLOG_ANNOUNCEMENT) {
```
Comment (RexXiong) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/FetchHandler.scala:92]: move this out from handleOpenStream method..

**Review by RexXiong — APPROVED** [review]

**Comment by RexXiong** [comment]
pls add more ut for Flink plugin.
